### PR TITLE
feat: add adjoint for `remake_buffer`

### PR DESCRIPTION
**Pull request by AayushSabharwal**
## Checklist

- [ ] Appropriate tests were added
- [ ] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [ ] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [ ] Any new documentation only uses public API

### CODE REVIEW

File: ext/MTKChainRulesCoreExt.jl [hunk 1]
```diff
@@ -2,14 +2,102 @@ module MTKChainRulesCoreExt
...
+
+    function remake_buffer_pullback(buf′)
+        buf′ = unthunk(buf′)
+        f′ = NoTangent()
+        indp′ = NoTangent()
+
+        tunable = selected_tangents(buf′.tunable, tunable_idxs)
+        discrete = selected_tangents(buf′.discrete, disc_idxs)
+        constant = selected_tangents(buf′.constant, const_idxs)
+        nonnumeric = selected_tangents(buf′.nonnumeric, nn_idxs)
```
Comment (ChrisRackauckas) [inline ext/MTKChainRulesCoreExt.jl:94]: is this type stable?

**Review by ChrisRackauckas — APPROVED** [review]

File: ext/MTKChainRulesCoreExt.jl (hunk 1 above)
Comment (AayushSabharwal) [reply ext/MTKChainRulesCoreExt.jl:94]: I haven't checked, though the type-stability of the adjoint will largely depend on the type stability of `idxs` and `vals`

Comment (AayushSabharwal) [reply ext/MTKChainRulesCoreExt.jl:94]: `selected_tangents` is type-stable, the pullback is almost type stable with a couple changes if `idxs` and `vals` are type-stable. The only source of instability is the one `selected_tangents` method which uses `similar(tangents, Union{T, NoTangent})` (which in fairness is a concretely typed union) and `vals′`

Comment (AayushSabharwal) [reply ext/MTKChainRulesCoreExt.jl:94]: However, if we restrict to the case of `idxs` only referring to values in the tunables, it's type-stable.

**Pull request merged**